stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,163 @@
+proc.shr_16
+    u32split
+    push.65536
+    mul
+    u32split
+    push.4294967296
+    mul
+    swap.2
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    add
+    add
+end
+
+#! Fetches the keyed value of a Sparse Merkle tree with the provided root.
+#!
+#! `V` will be `0` if there is no leaf value for `K` in `R`.
+#!
+#! input:  [K, R, ...]
+#! output: [V, R, ...]
+export.get
+    # invoke adv and fetch target depth
+    # [d, K, R, ...]
+    adv.smtget adv_push.1
+
+    # if has value, compute node value; use empty root otherwise
+    # [N, d, K, R, V, ...]
+    adv_push.1
+    if.true
+        # make a domain word for depth
+        # [0, 0, d, 0, K, R, ...]
+        push.0 swap push.0.0
+
+        # load the value
+        # [V, 0, 0, d, 0, K, R, V, ...]
+        padw adv_loadw movdnw.3 dupw.3
+
+        # compute remaining path key
+        # [K', V, 0, 0, d, 0, K, R, V, ...]
+        dupw.2 dup.10 eq.16
+        if.true
+            exec.shr_16
+        else
+            dup.10 eq.32
+            if.true
+                exec.shr_32
+            else
+                dup.10 eq.48
+                if.true
+                    exec.shr_48
+                else
+                    dup.10 eq.64
+                    if.true
+                        # remaining path of depth 64 is 0
+                        drop push.0
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
+
+        # perform the permutation and keep only digest output
+        # [N, d, K, R, V, ...]
+        swapw dup.10 movdn.12 hperm dropw swapw dropw
+    else
+        # [d, K, R, V, ...]
+        padw padw drop movdnw.3 push.0 dropw
+
+        # push empty constant as node
+        # [N, d, K, R, V, ...]
+        dup eq.16
+        if.true
+            push.17483286922353768131
+            push.353378057542380712
+            push.1935183237414585408
+            push.4820339620987989650
+        else
+            dup eq.32
+            if.true
+                push.11677748883385181208
+                push.15891398395707500576
+                push.3790704659934033620
+                push.2126099371106695189
+            else
+                dup eq.48
+                if.true
+                    push.10650694022550988030
+                    push.5634734408638476525
+                    push.9233115969432897632
+                    push.1437907447409278328
+                else
+                    dup eq.64
+                    if.true
+                        padw
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
```

Comment:
I think we might be able to simplify this and get rid of conditional statements by first writing the 4 possible node values to memory (using `adv_pipe` instruction) and then loading the appropriate value from memory.